src/main/java/org/testng/internal/MethodInvocationHelper.java

```diff
@@ -157,6 +157,8 @@ protected static Object invokeMethod(Method thisMethod, Object instance, Object[
         parameters.add(method);
       } else if (cls.equals(ITestContext.class)) {
         parameters.add(testContext);
+      } else if (cls.equals(Class.class)) {
+        parameters.add(com.getDeclaringClass());

```

Comment:
> Data providers are typically non static

@cbeust Could you elaborate why? I don't recommend non static data provider most of the time because I think it is better to separate the generation logic and the test logic.